modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -0,0 +1,480 @@
+
+        // Map IndexRow to Cache Key-Value pair.
+        return new GridCloseableIteratorAdapter<IgniteBiTuple<K, V>>() {
+            private IndexRow currVal;
+
+            private final CacheObjectContext coctx = cctx.cacheObjectContext();
+
+            /** {@inheritDoc} */
+            @Override protected boolean onHasNext() throws IgniteCheckedException {
+                if (currVal != null)
+                    return true;
+
+                if (!cursor.next())
+                    return false;
+
+                currVal = cursor.get();
+
+                return true;
+            }
+
+            /** {@inheritDoc} */
+            @Override protected IgniteBiTuple<K, V> onNext() {
+                if (currVal == null)
+                    if (!hasNext())
+                        throw new NoSuchElementException();
+
+                IndexRow row = currVal;
+
+                currVal = null;
+
+                K k = (K) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().key(), keepBinary, false);
+                V v = (V) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().value(), keepBinary, false);
+
+                return new IgniteBiTuple<>(k, v);
+            }
+        };
+    }
+
+    /** Get index to run query by specified description. */
+    private Index index(GridCacheContext cctx, IndexQueryDesc idxQryDesc) throws IgniteCheckedException {
+        Class<?> valCls = idxQryDesc.valCls() != null ? loadValClass(cctx, idxQryDesc.valCls()) : null;
+
+        String tableName = cctx.kernalContext().query().tableName(cctx.name(), valCls);
+
+        if (tableName == null)
+            return null;
+
+        // Find index by specified name.
+        if (idxQryDesc.idxName() != null) {
+            String name = "_key_PK".equals(idxQryDesc.idxName()) ? "_key_PK" : idxQryDesc.idxName().toUpperCase();
+
+            String schema = idxQryDesc.schema() == null ? cctx.name() : idxQryDesc.schema();
+
+            IndexName idxName = new IndexName(cctx.name(), schema, tableName, name);
+
+            Index idx = idxProc.index(idxName);
+
+            if (idx == null)
+                return null;
+
+            return checkIndex(idxProc.indexDefinition(idx.id()), idxQryDesc.idxCond()) ? idx : null;
+        }
+
+        // Try get index by list of fields to query.
+        // Check all indexes by cache to find best index match: count of index fields equals to count of index condition fields.
+        Collection<Index> idxs = idxProc.indexes(cctx);
+
+        Index idx = null;
+        int idxFieldsCnt = 0;
+
+        for (Index i: idxs) {
+            IndexDefinition idxDef = idxProc.indexDefinition(i.id());
+
+            if (!tableName.equals(idxDef.idxName().tableName()))
+                continue;
+
+            int fldsCnt = idxDef.indexKeyDefinitions().size();
+
+            if (checkIndex(idxDef, idxQryDesc.idxCond())) {
```

Comment:
It's trying to find best match in loop. For example there are 2 indexes (a, b, c) and (a, b). So fo queries (a, b) both indexes suite. But the second one is preferred.

Reply:
1. Indexes (a, b, c) and (a, b) have the same selectivity for (a, b) query. And without analyzing inline size it's impossible to say which one is preferred.
2. It's a strange case, because having both (a, b, c) and (a, b) indexes brings only performance drop, and (a, b) index should be dropped.